ckanext/cloudstorage/logic/action/multipart.py

```diff
@@ -0,0 +1,221 @@
+        except AttributeError:
+            upload_id_list = filter(
+                lambda e: e.tag.endswith('UploadId'),
+                resp.object.getchildren()
+            )
+            upload_id = upload_id_list[0].text
+        upload_object = MultipartUpload(upload_id, id, res_name, size, name)
+
+        upload_object.save()
+    return upload_object.as_dict()
+
+
+def upload_multipart(context, data_dict):
+    h.check_access('cloudstorage_upload_multipart', data_dict)
+    upload_id, part_number, part_content = toolkit.get_or_bust(
+        data_dict, ['uploadId', 'partNumber', 'upload'])
+
+    uploader = ResourceCloudStorage({})
+    upload = model.Session.query(MultipartUpload).get(upload_id)
+
+    resp = uploader.driver.connection.request(
+        _get_object_url(
+            uploader, upload.name) + '?partNumber={0}&uploadId={1}'.format(
+                part_number, upload_id),
+        method='PUT',
+        data=bytearray(part_content.file.read())
+    )
+    if resp.status != 200:
+        raise toolkit.ValidationError('Upload failed: part %s' % part_number)
+
+    _save_part_info(part_number, resp.headers['etag'], upload)
+    return {
+        'partNumber': part_number,
+        'ETag': resp.headers['etag']
+    }
+
+
+def finish_multipart(context, data_dict):
+    """Called after all parts had been uploaded.
+
+    Triggers call to `_commit_multipart` which will convert separate uploaded parts into single file
+
+    :param context:
+    :param data_dict:
+        dict with required key `uploadId` - id of Multipart Upload that should be finished
+    :returns: None
+    :rtype: NoneType
+
+    """
+
+    h.check_access('cloudstorage_finish_multipart', data_dict)
+    upload_id = toolkit.get_or_bust(data_dict, 'uploadId')
+    upload = model.Session.query(MultipartUpload).get(upload_id)
+    chunks = [
+        (part.n, part.etag)
+        for part in model.Session.query(MultipartPart).filter_by(
+            upload_id=upload_id).order_by(MultipartPart.n)
+    ]
+    uploader = ResourceCloudStorage({})
+    try:
+        obj = uploader.container.get_object(upload.name)
+        obj.delete()
+    except:
+        pass
+    uploader.driver._commit_multipart(
+        _get_object_url(uploader, upload.name),
+        upload_id,
+        chunks)
+    upload.delete()
+    upload.commit()
+
+    try:
+        res_dict = toolkit.get_action('resource_show')(
+            context.copy(), {'id': data_dict.get('id')})
+        pkg_dict = toolkit.get_action('package_show')(
+            context.copy(), {'id': res_dict['package_id']})
+
+        if pkg_dict['state'] == 'draft':
+            toolkit.get_action('package_patch')(
+                dict(context.copy(), allow_state_change=True),
+                dict(id=pkg_dict['id'], state='active')
+            )
+    except Exception as e:
+        log.error(type(e))
+        log.error(e)
+    h.flash_success('File successfully uploaded.')
```

Comment:
Setting flash messages from an action assumes the action is called from a controller not as an API call, so it's probably not a good idea. Is there a better way this message could be displayed?